SCAutolib/controller.py

```diff
@@ -221,136 +275,143 @@ def setup_user(self, user_dict: dict, force: bool = False):
                 new_user.delete_user()
-            user_dict["card_dir"].mkdir(exist_ok=True)
-
             new_user.add_user()
-
-        if user_dict["card_type"] == "virtual":
-            hsm_conf = file.SoftHSM2Conf(
-                new_user.card_dir.joinpath("sofhtsm2.conf"),
-                new_user.card_dir)
-            hsm_conf.create()
-            hsm_conf.save()
-
-            new_user.card = card.VirtualCard(new_user,
-                                             softhsm2_conf=hsm_conf.path)
-        else:
-            raise NotImplementedError("Other card type than 'virtual' does"
-                                      "not supported yet")
-
-        new_user.card.create()
         self.users.append(new_user)
-
         dump_to_json(new_user)
-
         return new_user
 
-    def enroll_card(self, user_: user.User, force: bool = False):
+    def setup_card(self, card_dict: dict, force: bool = False):
         """
-        Enroll the card of a given user with configured CA. If private key
-        and/or the certificate do not exists, new one's would be requested
-        from corresponding CA.
+        Create card object. Card object should contain its root CA cert as it
+        represents general card (i.e. including physical read-only cards).
 
-        :param force:
-        :param user_: User with a card to be enrolled.
+        :param card_dict: Dictionary containing card attributes
+        :type card_dict: dict
+        :param force: If its true and card directory exists it will be removed
+        :type force: bool
         """
-        logger.debug(f"Starting enrollment of the card for user "
-                     f"{user_.username}")
-        if not user_.card:
-            raise exceptions.SCAutolibException(
-                f"Card for the user {user_.username} is not initialized")
+        card_dir: Path = Path("/root/cards", card_dict["name"])
+        card_dir.mkdir(parents=True, exist_ok=True)
 
-        if not user_.key.exists():
-            _gen_private_key(user_.key)
+        if force and card_dir.exists():
+            rmtree(card_dir)
 
-        if not user_.cert.exists():
-            # Creating a new private key makes sense only if the certificate
-            # doesn't exist yet
+        if card_dict["card_type"] == CardType.physical:
+            new_card = card.PhysicalCard(card_dict, card_dir=card_dir)
+        elif card_dict["card_type"] == CardType.virtual:
+            hsm_conf = self.prepare_softhsm_config(card_dir)
+            new_card = card.VirtualCard(card_dict, softhsm2_conf=hsm_conf.path,
+                                        card_dir=card_dir)
+            # card needs to know some details of its user
+            new_card.user = self.link_user_to_card(new_card)
+            if new_card.user.user_type == UserType.local:
+                new_card.cnf = self.prepare_user_cnf(new_card)
+            if force:
+                self.revoke_certs(new_card)
+            new_card.create()
```

Comment:
I assume that in such case it will use existing certificates - i.e. it will not create new certs later on. This could be useful if for instance one wants to use specific certs. These certs will not be erased if force=False.